src/sqlfluff/rules/structure/ST04.py

```diff
@@ -104,36 +105,136 @@ def _eval(self, context: RuleContext) -> LintResult:
+            stop_seg=case1_else_expressions.get(),
+        )
 
         # Delete the nested "CASE" expression.
-        fixes = case1_to_delete.apply(lambda seg: LintFix.delete(seg))
+        fixes = case1_to_delete.apply(LintFix.delete)
 
         tab_space_size: int = context.config.get("tab_space_size", ["indentation"])
         indent_unit: str = context.config.get("indent_unit", ["indentation"])
 
         # Determine the indentation to use when we move the nested "WHEN"
         # and "ELSE" clauses, based on the indentation of case1_last_when.
         # If no whitespace segments found, use default indent.
-        indent = (
-            case1_children.select(stop_seg=case1_last_when)
-            .reversed()
-            .first(sp.is_type("whitespace"))
+        when_indent_str = self._get_indentation(
+            case1_children, case1_last_when, tab_space_size, indent_unit
         )
-        indent_str = (
-            "".join(seg.raw for seg in indent)
-            if indent
-            else construct_single_indent(indent_unit, tab_space_size)
+        # Again determine indentation, but matching the "CASE"/"END" level.
+        end_indent_str = self._get_indentation(
+            case1_children, case1_first_case, tab_space_size, indent_unit
         )
 
         # Move the nested "when" and "else" clauses after the last outer
         # "when".
-        nested_clauses = case2.children(sp.is_type("when_clause", "else_clause"))
-        create_after_last_when = nested_clauses.apply(
-            lambda seg: [NewlineSegment(), WhitespaceSegment(indent_str), seg]
+        nested_clauses = case2.children(
+            sp.is_type("when_clause", "else_clause", "newline", "comment", "whitespace")
         )
-        segments = [item for sublist in create_after_last_when for item in sublist]
+
+        # Rebuild the nested case statement.
+        # Any comments after the last outer "WHEN" that were deleted
+        segments = list(case1_comments_to_restore)
+        # Any comments between the "ELSE" and nested "CASE"
+        segments += self._rebuild_spacing(when_indent_str, after_else_comment)
+        # The nested "WHEN", "ELSE" or "comments", with logical spacing
+        segments += self._rebuild_spacing(when_indent_str, nested_clauses)
         fixes.append(LintFix.create_after(case1_last_when, segments, source=segments))
 
         # Delete the outer "else" clause.
         fixes.append(LintFix.delete(case1_else_clause_seg))
+        # Add spacing for any comments that may exist after the nested `END`
+        # but only on that same line.
+        fixes += self._nested_end_trailing_comment(
+            case1_children, case1_else_clause_seg, end_indent_str
+        )
         return LintResult(case2[0], fixes=fixes)
+
+    def _get_indentation(
+        self,
+        parent_segments: Segments,
+        segment: BaseSegment,
+        tab_space_size: int,
+        indent_unit: str,
+    ):
```

Comment:
Can you add the return type for these new functions?

Reply:
Added